% !TeX encoding = UTF-8
% !TeX program = pdflatex
\documentclass[english, noexaminfo, a4paper, binding=0.6cm, oneside]{sapthesis}

\usepackage[utf8]{inputenc}
\usepackage{indentfirst}
\usepackage{microtype}
\usepackage{colortbl}
\usepackage[table]{xcolor}
\usepackage{booktabs}
\usepackage{lettrine}
\usepackage{graphicx}
\usepackage{subcaption}
\usepackage[nottoc, notlof, notlot]{tocbibind}
\usepackage{microtype}
\usepackage[english]{babel}
\usepackage[utf8]{inputenc}
\usepackage{listings}
\usepackage{xcolor}
\usepackage{amsfonts}
\usepackage[hidelinks]{hyperref}
\usepackage[nameinlink]{cleveref}
\usepackage{hyperref}
\usepackage{algorithmicx}
\usepackage[noend]{algpseudocode}
\usepackage{algorithm}
\usepackage{biblatex}
\usepackage{graphicx}
\usepackage{tikz}

\usepackage{amsmath}
\DeclareMathOperator*{\argmin}{argmin}

\definecolor{lightgray}{gray}{0.9} % 0.9 is very light, 0.1 is very dark
% \linespread{0.9}
\hypersetup{
    hyperfootnotes=true,			
    bookmarks=true,			
    colorlinks=true,
    linkcolor=red,
    linktoc=page,
    anchorcolor=black,
    citecolor=red,
    urlcolor=blue,
    pdftitle={AI-driven analysis of molecular pathways},
    pdfauthor={Ionuț Cicio},
    pdfkeywords={thesis, sapienza, roma, university}
}

\title{AI-driven analysis of molecular pathways}
\author{Ionuț Cicio}
\IDnumber{2048752}
\course{Informatica}
\courseorganizer{Facolt\`a di Ingegneria dell'informazione, informatica e statistica}
\submitdate{2024/2025}
\copyyear{2025}
\advisor{Prof. Mancini}
\coadvisor{Prof. Tronci}
\authoremail{cicio.2048752@studenti.uniroma1.it}
%\examdate{xx December 2025}
%\examiner{Prof. ...} \examiner{Prof. ...} \examiner{Prof. ...}  \examiner{Prof. ...} % \examiner{Prof. ...} \examiner{Prof. ...}  \examiner{Prof. ...} 

%we refer to http://ctan.mirrorcatalogs.com/macros/latex/contrib/sapthesis/sapthesis-doc.pdf for an exhaustive description of the sapthesis documentclass.
\definecolor{codegreen}{rgb}{0,0.6,0}
\definecolor{codegray}{rgb}{0.5,0.5,0.5}
\definecolor{codepurple}{rgb}{0.58,0,0.82}
\definecolor{backcolour}{rgb}{0.95,0.95,0.92}
% --- Doom One Light Color Definitions ---
% (Add this to your preamble)
\definecolor{doomBack}{RGB}{250, 250, 250}     % #fafafa
\definecolor{doomComment}{RGB}{156, 156, 156}   % #9c9c9c
\definecolor{doomKeyword}{RGB}{198, 120, 221}   % #c678dd (Purple/Magenta)
\definecolor{doomString}{RGB}{152, 195, 121}    % #98c379 (Green)
\definecolor{doomNumber}{RGB}{209, 154, 102}    % #d19a66 (Orange)
\definecolor{doomDefault}{RGB}{77, 77, 77}      % #4d4d4d (Default Text)
% Setup section. -----------------------------------|
\crefname{listing}{listato}{listati}
\Crefname{listing}{Listato}{Listati}
\crefname{lstlisting}{listato}{listati}
\Crefname{lstlisting}{Listato}{Listati}

\crefname{section}{sezione}{sezioni}
\Crefname{section}{Sezione}{Sezioni}
\crefname{chapter}{capitolo}{capitoli}
\Crefname{chapter}{Capitolo}{Capitoli}
\lstset{
basicstyle= % <-- Add this for the rest of the text
}
\lstdefinestyle{pythonstyle}{
  language=Python,
  basicstyle=\ttfamily\small,
  backgroundcolor=\color{doomBack}, 
  commentstyle=\color{doomComment},
  keywordstyle=\color{doomKeyword},
  numberstyle=\tiny\color{doomNumber},
  stringstyle=\color{doomString},
  basicstyle=\ttfamily\footnotesize\color{doomDefault},
  breakatwhitespace=false,         
  breaklines=true,                 
  captionpos=b,                    
  keepspaces=true,                 
  numbers=left,                    
  numbersep=8pt,                  
  showspaces=false,                
  showstringspaces=false,
  showtabs=false,                  
  tabsize=2
}

% Stile terminale (bash/sh)
\lstdefinestyle{termstyle}{
  language=bash,
  basicstyle=\ttfamily\small\color{codegreen},
  backgroundcolor=\color{backcolour},
  showstringspaces=false,
  breaklines=true,
  columns=fullflexible,
  keepspaces=true,
  upquote=true,
  frame=single,
  rulecolor=\color{black},
  numbersep=8pt,
  numbers=none,
  captionpos=b
}
% Some typical keyword
\lstdefinelanguage{mybash}[]{bash}{
  morekeywords={sudo,pip,python,python3,git,make,cmake,virtualenv,conda}
}
\lstdefinestyle{termstyle}{
  language=mybash,
  basicstyle=\ttfamily\small\color{codegreen},
  backgroundcolor=\color{backcolour},
  showstringspaces=false, breaklines=true,
  columns=fullflexible, keepspaces=true, upquote=true,
  frame=single, rulecolor=\color{black}, numbers=none, captionpos=b
}

\addbibresource{bibliography.bib}

\newcounter{definition}
\renewcommand{\thedefinition}{\arabic{definition}}

\newenvironment{definition}[1][]{
    \refstepcounter{definition}
    \par\noindent\linebreak
    \textbf{Definition \thedefinition } (#1).}{
    \par\noindent
}

\begin{document}

\frontmatter

\maketitle

\begin{abstract} 

    In this report it's presented a tool that uses molecular pathway databases
    to generate biochemical networks by computing the backwards-reachable
    reactions from a set of species of interest. The biochemical network is
    then extended with a system of ODE which describes its dynamics  by using
    mass action kinetics.

%     Model-based formal verification of industry-relevant Cyber-Physical Systems (CPSs) is often a computationally
% prohibitive task. In most cases, the complexity of the models precludes any prospect of symbolic analysis,
% leaving numerical simulation as the only viable option. Unfortunately, exhaustive simulation of a CPS model
% over the entire set of plausible operational scenarios is rarely possible in practice, and alternative strategies
% such as Statistical Model Checking (SMC) must be used instead.
% In this article, we show that the number of model simulations (samples) required by SMC techniques to
% converge can be significantly reduced by considering multiple (an ensemble of) Adaptive Stopping Algorithms
% (SAs) at once, and that the simulations themselves (by far the most expensive step of the entire workload) can
% be efficiently sped up by exploiting massively parallel platforms.
% With three industry-scale CPS models, we experimentally show that the use of an ensemble of two state-of-
% the-art SAs (AA and EBGStop) may require dozens of millions fewer samples when compared to running a single
% algorithm, with reductions in sample size of up to 78\%. Furthermore, we show that our implementation, by
% massively parallelizing system model simulations on a HPC infrastructure, yields speed-ups for the completion
% time of the verification tasks which are practically linear with respect to the number of computational nodes,
% thus achieving an efficiency of virtually 100\%, even on very large platforms. This makes it possible to complete
% tasks of model-based SMC verification for complex CPSs in a matter of hours or days, whereas a naïve
% sequential execution would require from months to many years.

\end{abstract}

\tableofcontents

\mainmatter

\chapter{Introduction}

Biological systems such as organisms, cells, or biomolecules are highly
organized in their structure and function, \cite{Klipp16}. Such structures can
help study the and compare biological functions of different organisms. In
order to understand, formalize, abstract and make predictions on these
structures a mathematical model of these structures is needed.

Given a set of biochemical species and metabolic pathways of interest, this
report presents a tool which generates biological models by using the Reactome
pathway database \cite{Fabregat2018} to compute the biochemical network of all
the backwards-reachable species and reactions starting from the species of
interest and within the metabolic pathways of interest. 

The generated biochemical networks are qualitative models and contain
information about the backwards-reachable reactions (and reaction-like events),
their products, reactants (with the stoichiometries), modifiers and
compartments. In order to study the dynamic behaviour of these biochemical
networks a system of ODE is produced by using the Law of Mass Action to
describe the rates of the reactions within the network, and new reactions are
added to produce and consume the species on the boundary of the network.

The product is a SBML (Systems Biology Markup Language) \cite{sbml} file with
the dynamics of the biochemical network and a set of decision variables that
determine the speeds of the reactions within the network. This model is then
subject to a study of stability by defining a set of constraints for stability
and using the OpenBox BBO (blackbox optimization) solver \cite{li2021openbox}
to find \textit{virtual patients} (assignements to the decision variables)
which produce stable states for the model. 

Validating the stability constraints for a \textit{virtual} patient requries
simulating the biological model, which is a very computationally expensive task
when done multiple times; in order to reduce steady states search times an
orchestrator-worker architecture is introduced by using the OpenBox as a
Service \cite{OpenBoxServiceDocumentation} feature.


% [A fundamental component of bioinformatics is data integration, (a problem is)
% i.e. partial information in distributed databases is needed]. [One such data
% source is Reactome \cite{Fabregat2018}, which is a qualitative network database]
% REACTOME is an open-source, open access, manually curated and peer-reviewed
% pathway database. [The goal with this work is to use qualitative data in the
% Reactome database to generate quantitative models, and use BBO (cite something)
% techinques to "validate" (validate is not good, find something else) these
% models]
%
% [Validating complex biological models is a computationally intensive task (cite
% something) thus HPC clusters are required in order. ] {Such clusters may not
% always be fully available (i.e. there are multiple users and multiple
% experiments running on the cluster + there are some limits on job running
% times), so in order to better distribute the load of the validation task some
% infrastructure work is needed.}

\newpage


\section{Preliminaries}

% We denote by $\mathbb{R}, \mathbb{R}_{0_{+}}, \mathbb{Z}, \mathbb{N}$ the sets of, respectively, real, non-
% negative real, integer, and non-negative integer numbers.
The set of non-zero natural numbers is denoted with $\mathbb{N}_+ = \mathbb{N} - \{ 0 \}$.

This report uses the terms \textit{species} and \textit{physical entity}
interchangibly to refer to biochemical species, since Reactome
\cite{ReactomeDataModelGlossary} uses the term \textit{PhysicalEntity} and SBML
\cite{sbmlLevel3Version2CoreRelease2} uses the term \textit{species}. 

Reactome uses the term \textit{ReactionLikeEvent} to refer to reactions
\cite{ReactomeDataModelGlossary}, since it encompasses different types of
events other than simple chemical reactions (such as translocation of molecules
from one compartment to another, association of molecules to form complexes and
polymers, dissociation of complexes and polymers etc...).


% (TODO: cite glossary and use glossary to tell different types of reactions)

% (because X, Y, the two main technologies used in this work, use the
% two terms) r

% - TODO: what is a species
%
% - TODO: what is a reaction
%
% - TODO: what is a compartment (use footnotes)


\section{Biochemical networks} % TODO: Motivation

Biochemical networks are one of the main concepts used to model biological
systems \cite{Koutrouli2023erratum}. Such networks are made up of relations
between physical entities (small molecules, proteins, nucleic acids, chemical
compounds, complexes, larger macromolecular assemblies, atoms, electrons, and
photons \cite{ReactomeDataModelGlossary}) and reactions (such as standard
chemical reactions, translocation of molecules from one compartment to another,
association of molecules to form complexes and polymers, etc...
\cite{ReactomeDataModelGlossary}). More often than not biochemical networks are
organized in pathways, sets of causally connected reactions
\cite{ReactomeDataModelGlossary}; such pathways, in pathway databases like
Reactome (Figure \ref{fig:reactome-uml}) are organized hierarchically: Reactome
has top-level pathways, like \textit{Signal transduction} (reactions where
extracellular signals elicit changes in cell state and activity),
\textit{Muscle contraction} (Figure
\ref{fig:striated-muscle-contraction-hierarchy}) and the \textit{Neuronal
system}, each composed of other smaller pathways or reactions.

%TODO: cite signal transduction definition from reactome

\begin{figure}[H]
    \includegraphics[width=\textwidth]{./images/striated_muscle_contration.pdf}
    % \includegraphics[]{./images/legend.png}
    % (R-HSA-390522.3) 
    % https://reactome.org/PathwayBrowser/#/R-HSA-390522&PATH=R-HSA-397014
    \caption{
        \textit{Striated muscle contraction} pathway within the \textit{Muscle
        contraction} pathway in the Reactome PathwayBrowser
        \cite{ReactomeStriatedMuscleContraction}
    }
    \label{fig:striated-muscle-contraction}
\end{figure}

% TODO: add legend here, or somewhere

% Here striated muscle contraction is represented on the basis of calcium binding
% to the troponin complex, which exposes the active sites of actin. Once the
% active sites of actin are exposed, the myosin complex bound to ADP can bind
% actin and the myosin head can pivot, pulling the thin actin and thick myosin
% filaments past one another. Once the myosin head pivots, ADP is ejected, a fresh
% ATP can be bound and the energy from the hydrolysis of ATP to ADP is channeled
% into kinetic energy by resetting the myosin head. With repeated rounds of this
% cycle the sarcomere containing the thin and thick filaments effectively
% shortens, forming the basis of muscle contraction.

The biochemical network in Figure \ref{fig:striated-muscle-contraction}
represents the \textit{Striated muscle contraction} pathway in the
\textit{Cytosol} compartment, a process where force is generated within
striated muscle tissue, resulting in muscle movement
\cite{ReactomeStriatedMuscleContraction}. Here the pathway has the following
reactions:

% association / binding

% Striated muscle contraction Figure  is a process
% whereby force is generated within striated muscle tissue, resulting in a change
% in muscle geometry, or in short, increased force being exerted on the tendons.
% Force generation involves a chemo-mechanical energy conversion step that is
% carried out by the actin/myosin complex activity, which generates force through
% ATP hydrolysis. Striated muscle is a type of muscle composed of myofibrils,
% containing repeating units called sarcomeres, in which the contractile
% myofibrils are arranged in parallel to the axis of the cell, resulting in
% transverse or oblique striations observable at the level of the light
% microscope.

\begin{enumerate}
    \item Ca$^{2_+}$ (stoichiometry 2) binds with an \textit{Inactive Sarcomere
        Protein Complex}
    \item ATP (stoichiometry 2) associates with the bound "Ca$^{2_+}$ - \textit{Calcium
        Sarcomere Protein Complex}"
    \item The \textit{Myosyn Complex} acts as a catalyst in order to activate the
        reaction that produces \textit{Inorganic Phosphate} and the "ADP - \textit{Calcium
        Sarcomere Protein Complex}" bound
    \item The ADP dissociates from the "Ca$^{2_+}$ - \textit{Sarcomere Protein Complex}"
        bound, producing two units of ADP (stoichiometry 2)
\end{enumerate}

In particular, Ca$^{2_+}$, ATP, \textit{Inactive Sarcomere Protein Complex} and
\textit{Myosyn Complex}  are inputs of the biochemical network (no reaction
within the network produces these species), while Pi and ADP are outputs (no
reaction within the network consumes these species).

It's important to note that since the \textit{Myosyn Complex} is a catalyst
(modifier) of the reaction, it is not consumed in the reaction.

Biochemical networks as presented in pathway databases are qualitative models,
they describe only the roles of species in reactions and the structure of the
network. Other than the stoichiometries there are no informations about typical
quantities of species, reactions speeds, compartments sizes etc... 


\begin{definition}[Biochemical network] 
    A biochemical network $G$ is a tuple $(S, R, E, \sigma)$ where 
    \begin{itemize}
        % TODO: possibly not disjoint
        \item $S = U \cup X \cup Y$ is the set of species of the biochemical
            network, where $U, X, Y$ are possibly non disjoint sets s.t.
            \begin{itemize}
                \item $U$ is the set of input species of the network
                \item $X$ is the set of species which are neither inputs or outputs 
                \item $Y$ is the set of output species of the network
            \end{itemize}
        \item $R$ is the set of reactions in the biochemical network

        \item $T = \{ \text{reactant}, \text{product}, \text{modifier}^+,
            \text{modifier}^- \}$ is the set of relationship types between
            species and reactions

        \item $E \subseteq S \times R \times T$ is the set of relationships between 
            species and reactions

        \item $E_t = \{ (s, r) \; | \; (s, r, t) \in E \}$ is the projection of $E$ over $t$

        % TODO: E \subseteq S times R times T where T is the type 
        % \item $E = E_\text{reactants} \cup E_\text{products} \cup
        %     E_\text{modifiers}^+ \cup E_\text{modifiers}^- \subseteq S
        %     \times R$ is the set of relations between species and reactions

            % TODO: rewrite
        % \item $E = (E_\text{reactants}, E_\text{products},
        %     E_\text{modifiers}^+,  E_\text{modifiers}^-)$ is a tuple with the
        %     sets of relations between species and reactions with $E_i \subseteq
        %     S \times R$
        \item $\sigma : E_\text{reactant} \cup E_\text{product} \rightarrow
            \mathbb{N}_+$ is the stoichiometry of the reactants of products of
            the reaction
    \end{itemize}
\end{definition}

The definition abstracts different reaction types. It also allows for a species to be
have multiple roles in a reaction, since in the Reactome pathways database there 
are some reactions in which, for example, modifiers are also inputs or outputs of the
reaction. (TODO: reference cypher query or species)

% TODO: Nell'esempio precedente S, R, E sono etc...
% TODO: rclone drive 


% \lstset{
%     language=cypher,
%     basicstyle=\ttfamily\small,
%     keywordstyle=\color{blue},
%     commentstyle=\itshape\color{green},
%     stringstyle=\color{red},
%     breaklines=true,
%     numbers=left,
%     numbersep=5pt,
%     tabsize=2,
%     frame=single
% }
%
% \begin{lstlisting}[caption=Cypher Query Example, language=cypher]
% MATCH (n:Person {name: "Alice"}) 
% RETURN n.name, n.age
% \end{lstlisting}

\newpage

\section{Dynamics of biochemical networks}

Qualitative models, such as the aforementioned biochemical networks, are not
enough to study the behaviour of a pathway, a description of how the network
evolves over time when put in a stable environment is needed in order to make
predictions and study what happens when the model is subject to perturbations.
Systems of ordinary differential equations (ODE) are one of the approaches used
in order to describe the dynamic behaviour of biochemical networks
\cite{Klipp16}.

\begin{figure}[H]
    \begin{center}
        \includegraphics[width=.6\textwidth]{./images/hemoglobin.pdf}
    \end{center}
    \caption{
        \textit{Hemoglobin A} binds O$_2$, releasing H$^+$ and CO$_2$, then
        AQP1 acts as a catalyst and transports CO$_2$ from the cytosol to the
        extracellular region (reaction in the \textit{O$_2$/CO$_2$ exchange in erythrocytes} pathway \cite{ReactomePathwayBrowserOxigenExchagnePathway})
    }
    % TODO: species are different based on region
    % TODO: define what cytosol is, maybe with a footnote
    % TODO: recall to appendix to explain why physical entities have different regions
    \label{fig:hemoglobin}
\end{figure}

 Let $G = (S, R, E, \sigma)$ be the biochemical network in Figure \ref{fig:hemoglobin}

% TODO:  https://reactome.org/PathwayBrowser/#/R-HSA-1247673&PATH=R-HSA-382551,R-HSA-1480926
% https://reactome.org/PathwayBrowser/#/R-HSA-1247673&SEL=R-HSA-1247668&PATH=R-HSA-382551,R-HSA-1480926

\begin{itemize}
    \item $S = U \cup X \cup Y$ where
        \begin{itemize}
            \item $U = \{\text{O}_2, \text{Hb}_1, \text{A}\}$ with Hb$_1$ and A being
                respectively \textit{Hemoglobin A} before the dissociation and AQP1
            \item $X = \{ \text{CO}_2^c \}$, with $\text{CO}_2^c$ being
                $\text{CO}_2$ in the cytosol
            \item $Y = \{ \text{Hb}_2, H^+, \text{CO}_2^e \}$ with
                $\text{Hb}_2$ and $\text{CO}_2^e$ being respectively
                \textit{Hemoglobin A} after the dissociation and $\text{CO}_2$
                in the extracellular region
        \end{itemize}

    \item $R = \{r_1, r_2\}$ where $r_1$ and $r_2$ are respectively the
        dissociation reaction and the transport reaction
\end{itemize}

One possible way to describe the dynamics of $G$ is through the system of
ordinary differential equations described in Equations \ref{eq:ode-1},
\ref{eq:ode-2} and \ref{eq:ode-3}, where the rate of change of a given species
$S_i \in S$ depends on the rate of change of the reactions that produce and
consume it.

\newcommand{\dd}{\mathop{}\!\mathrm{d}}

\begin{equation}
    \begin{gathered}
        \frac{\dd \text{O}_2}{\dd t} = -r_1, \quad
        \frac{\dd \text{Hb}_1}{\dd t} = -r_1, \quad
        \frac{\dd \text{Hb}_2}{\dd t} = +r_1, \quad
        \frac{\dd \text{H}^+}{\dd t} = +r_1 \\
        \frac{\dd \text{CO}_2^c}{\dd t} = +r_1 - r_2, \quad
        \frac{\dd \text{CO}_2^e}{\dd t} = -r_1 + r_2, \quad
        \frac{\dd \text{A}}{\dd t} = 0 \\
    \end{gathered}
    \label{eq:ode-1}
\end{equation}

\begin{equation}
    r_1 = k_1 \cdot (\text{O}_2)^4 \cdot (\text{Hb}_1)^1
    \label{eq:ode-2}
\end{equation}
\begin{equation}
    r_2 = k_2 \cdot (\text{CO}_2^c)^4 \cdot \frac{A^{n_1}}{K_1 + A^{n_1}}
    \label{eq:ode-3}
\end{equation}

The reaction rate of $r_1$ and $r_2$ is given by mass action kinetics, where
the speed of reaction $r_i \in R$ is determined by the kinetic constant $k_i$
and is proportional to the concentrations of the reactants and modifiers. In
particular, since AQP1 acts as a catalyst in Equation \ref{eq:ode-3}, reaction
$r_2$ can be activated only if $A$ reaches a certain concentration which
depends on the constants $K_1$ and $n_1$ in the hill function.

\subsection{The Law of Mass Action}

The kinetics in Equantions \ref{eq:ode-2} and \ref{eq:ode-3} are based on the
mass action law, introduced by Guldberg and Waage in 1864
\cite{GuldbergWaage64}. It states that the reaction rate is proportional to the
probability of a collision of the reactants, probability which is in turn
proportional to the concentration of reactants to the power of the molecularity
(stoichiometry), which is the number in which the molecule species enter the
reaction (this is because the higher is the  number of molecule species need
for the reaction to happen, the lower is the probability the reaction will
happen) \cite{Klipp16}.  Given a biochemical network $G = (S, R, E, \sigma)$, and a
reaction $r \in R$, where $r$ is a \textbf{reversible} reaction, the rate
of the reaction $r$ in its most general form is written as 

\begin{equation}
    r = k_r^+ \cdot \prod_{\substack{(s, r) \\ \in \\ E_\text{reactants}}} s^{\sigma(s, r)} - 
    k_r^- \cdot \prod_{\substack{(s, r) \\ \in \\ E_\text{products}}} s^{\sigma(s, r)}
    \label{eq:law-of-mass-action}
\end{equation}

where

\begin{itemize}
    \item $s$ indicates the concentration of the species $s$
    \item $k_r^+$ is the kinetic constant of the forward reaction
    \item $k_r^-$ is the kinetic constant of the backward reaction
\end{itemize}

The form used for the generation of biological models in this report is the simpler one in Equation
\ref{eq:law-of-mass-action-simple}. This is because Reactome models reversible
reactions as two different reactions, so it's enough to include both the
forward and the backward reaction in order to obtain the equation for a single
reversible reaction.

\begin{equation}
    r = k_r \cdot \prod_{\substack{(s, r) \\ \in \\ E_\text{reactants}}} s^{\sigma(s, r)}
    \label{eq:law-of-mass-action-simple}
\end{equation}

The law of mass action has the advantage that a single decision variable is
associated to each reaction in order to determine the speed of the reaction.
This is a gross semplification since the behaviour of reactions is more complex
than this, but more detailed kinetics such as convenience kinetics introduce
many more decision variables, making the search of \textit{virtual patients}
for the model harder.


\subsection{Hill equation}

In nature the speed of reactions is heavily influenced by modifiers (enzymes
which speed-up the reaction and inhibitors which slow down or stop the
reaction). The characteristic of modifiers is that they are neither consumed or
produced by the reaction. Given a biochemical network $G = (S, R, E, \sigma)$, and a
pair $(s, r) \in E_\text{modifiers}^+ \cup E_\text{modifiers}^-$ (meaning
that the species $s$ is a modifier for reaction $r$, either as a catalyzer
or an inhibitor), the hill function can be defined as

\begin{equation}
    H_r^s =
    \begin{cases}
        \frac{(s)^{n_r^s}}{K_r^s + (s)^{n_r^s}}, \quad \text{if} \; (s, r) \in E_\text{modifiers}^+ \\[2ex]
        \frac{K_r^s}{K_r^s + (s)^{n_r^s}}, \quad \text{if} \; (s, r) \in E_\text{modifiers}^- \\
    \end{cases}
\end{equation}

where 

\begin{itemize}
    \item $K_r^s$ is the apparent dissociation constant of modifier $s$ in
        reaction $r$
    \item $n_r^s$ the hill coefficient of modifier $s$ (often in literature
        it's indicated with $h$) % TODO: citation
\end{itemize}


\begin{figure}[H]
    \begin{tabular}{ccc}
        \begin{tikzpicture}[scale=2.5]
            \draw[->,line width=1pt] (-0.25,0)--(1.25,0) node[right]{$x$};
            \foreach \x in {1}
                \draw (\x,0) node[below]{\x};

            \draw[->,line width=1pt] (0,-0.25)--(0,1.25) node[above]{$y$};
            \foreach \y in {1}
                \draw (0,\y) node[left]{\y};

            \draw[gray,dashed,very thin] (-0.25,-0.25) grid (1.25, 1.25);

            \draw[blue,line width=1pt,domain=0:1.25, samples=100] 
                plot(\x,{pow(\x,3)/(0.01+pow(\x,3))});
        \end{tikzpicture}
        \begin{tikzpicture}[scale=2.5]
            \draw[->,line width=1pt] (-0.25,0)--(1.25,0) node[right]{$x$};
            \foreach \x in {1}
                \draw (\x,0) node[below]{\x};

            \draw[->,line width=1pt] (0,-0.25)--(0,1.25) node[above]{$y$};
            \foreach \y in {1}
                \draw (0,\y) node[left]{\y};

            \draw[gray,dashed,very thin] (-0.25,-0.25) grid (1.25, 1.25);

            \draw[blue,line width=1pt,domain=0:1.25, samples=100] 
                plot(\x,{pow(\x,20)/(0.01+pow(\x,20))});
        \end{tikzpicture}
        \begin{tikzpicture}[scale=2.5]
            \draw[->,line width=1pt] (-0.25,0)--(1.25,0) node[right]{$x$};
            \foreach \x in {1}
                \draw (\x,0) node[below]{\x};

            \draw[->,line width=1pt] (0,-0.25)--(0,1.25) node[above]{$y$};
            \foreach \y in {1}
                \draw (0,\y) node[left]{\y};

            \draw[gray,dashed,very thin] (-0.25,-0.25) grid (1.25, 1.25);

            \draw[blue,line width=1pt,domain=0:1.25, samples=100] 
                plot(\x,{pow(\x,10)/(0.1+pow(\x,10))});
        \end{tikzpicture}
    \end{tabular}
    \caption{
        hill function for a catalyst with parameters $K_r^s, n_r^s$ respectively
        $(0.01, 3), (0.01, 10), (0.1, 10)$
    }
\end{figure}

With this information the kinetics of Equation \ref{eq:ode-3}, which describes
the transport of CO$_2$ from the cytosol to the extracellular region, can be
explained like this:

\begin{itemize}
    \item $k_1$ is the kinetic constant of the forward reaction
    \item $(\text{O}_2)^4$ is the concentration of O$_2$, raised to the power
        of 4 (the stoichiometry), given by the law of mass action
    \item $A$ acts as a catalyst for the transportation, so a hill function is
        used to moderate the speed of the reaction based on the concentration
        of $A$ 
\end{itemize}

% TODO: multiplication with law of mass action (somewhat done with example)

\newpage

In order to define the dynamics of a given biochemical network, the network
needs to be expanded to handle boundary species. In particular, for all species
that that are only inputs of the network a reaction needs to be added in order
to constantly produce the species (as the network needs to be simulated in a
stable environment). Contrarywise, all species which are only outputs of the
network need to be consumed. Some species are both inputs and outputs of the
network, these species are modifiers (otherwise they would either be consumed
in a reaction or produced by a reaction), and can be set to a fixed
concentration.

\begin{definition}[Biological model]
    Given a biochemical network $G = (S, R, E, \sigma)$ let $B = (G', \cal{K})$ be
    the biological model derived from $G$ with added mass action kinetics, with
    $G' = (S, R', E', s')$ where 
    \begin{itemize}
        \item $R' = R \cup R_\text{inputs} \cup R_\text{outputs}$ where
        \begin{itemize}
            \item $R_\text{inputs} = \{ r_{s} \; | \; s \in U - Y \}$ 
            \item $R_\text{outputs} = \{ r_{s} \; | \; s \in Y - U \}$ 
        \end{itemize}
        \item $E' = (E'_\text{reactants}, E'_\text{products}, E_\text{modifiers}^+,  E_\text{modifiers}^-)$, where
        \begin{itemize}
            \item $E'_\text{reactants} = E_\text{reactants} \cup \{ (s, r) \; | \; r_s \in R_\text{inputs} \}$ 
            \item $E'_\text{products} = E_\text{products} \cup \{ (s, r) \; | \; r_s \in R_\text{outputs} \}$
        \end{itemize}
% \{ (s, r) \; | \; r_s \in r_\text{inputs} \} \cup \{ (s, r) \; | \; r_s \in r_\text{outputs}\} E
    \end{itemize}
    \begin{equation}
        \begin{aligned}
        \end{aligned}
    \end{equation}

            % \begin{itemize}
            %     \item $E'_\text{reactants} = $
            % \item $E'_\text{products} = $
            % \end{itemize}

    Then $\cal{K}$ can be defined on $G'$

    \begin{equation}
        \begin{aligned}
            \cal{K} = & \{ k_r \; | \; r \in R' \} \; \cup \\
                & \{ K_r^s \; | \; r \in R' \land (s, r) \in E^+_\text{modifiers} \cup E^-_\text{modifiers} \} \; \cup \\
                & \{ n_r^s \; | \; r \in R' \land (s, r) \in E^+_\text{modifiers} \cup E^-_\text{modifiers} \} \; \cup
        \end{aligned}
    \end{equation}
    where 
    \begin{itemize}
        \item $k_r$ is the kinetic constant of reaction $r$
        \item $K_r^s$ is the apparent dissociation constant of modifier $s$ in reaction $r$
        \item $n_r^s$ the hill coefficient of modifier $s$ in reaction $r$
    \end{itemize}
\end{definition}

% TODO: apparent dissociation constant
% TODO: hill coefficient
% TODO: which dictates ...)


% Systems of ordinary differential equations (ODE) are
% probably the most frequently used approach to model
% the static and dynamic behaviors of biochemical net­
% works.
% They employ continuous variable values
% (mostly concentrations) and continuous time.

% \section{Qualitative network models augmentation} % TODO: Contribution
% \section{Qualitative network models}
% TODO: here cite reactome, put in "notes" the UML used later for the query etc...
% \section{Outline} % TODO: Outline
% TODO: rename in somthing like "finding virtual patients for a model"

% \begin{definition}[Biological network] A biological network $G$ is a tuple $(S, R, E, \sigma)$ where
% \begin{itemize}
%     \item $S = U \cup X \cup Y$ is the set of species of the biological network, where
%         \begin{itemize}
%             \item $U$ is the set of input species
%             \item $Y$ is the set of output species
%             \item $X$ is the set of other species in the network
%         \end{itemize}
%     \item $R$ is the set of reactions in the biological network
%     \item $E \subseteq S \times R$
% \end{itemize}
% \end{definition}


\chapter{Biological models generation}

% \chapter{Quantitative model generation}

% TODO: number of species, reactions and parameters in the network

% Given a set of target species, a set of constraints on the target species
% (constraints which model a scenario that could present, for example, in a
% disease) and by taking into account all the reactions within a set target
% pathways that lead to the production, both directly and indirectly, of the
% target species, the goal is to find a subset of virtual patients for the
% described scenario.

\section{Reactome and reachability}

REACTOME is an open-source, open access, manually curated and peer-reviewed
pathway database. (TODO: cite Reactome)


Given a biochemical species $S$ one possibile way to


\section{Scenarios}

\noindent\makebox[\textwidth]{
    \begin{tabular}{ccc}
    \includegraphics[width=60mm]{./images/nitric_oxide_signal_transduction_reactions.png} &
    \includegraphics[width=60mm]{./images/nitric_oxide_signal_transduction_entities.png} &
    \includegraphics[width=60mm]{./images/nitric_oxide_signal_transduction_time.png} \\ 
    \includegraphics[width=60mm]{./images/co_2_cytosol_molecules_transport_reactions.png} &
    \includegraphics[width=60mm]{./images/co_2_cytosol_molecules_transport_entities.png} &
    \includegraphics[width=60mm]{./images/co_2_cytosol_molecules_transport_time.png} 
    \end{tabular}
}

% (TODO: define what is an expansions, why do we need a scenario, etc...it might be important to study what a subsection which contains two specific species behaves)
%
% (TODO: define what is a Pathway, and what a Pathway is in terms of Reactome)
%
% (TODO: maybe do a chapter about Reactome, or something simpler before about the generation, maybe about "Reachability", for 
% reachability you need a definition of a network)


% // Average quantities
% //
% // - $S' = S union { S_"avg" | s in S}$
% // - $S' = G(S')$
% // - $K: R -> RR_+^(|R|) = [10^(-6), 10^6]^(|R|)$
% //
% // // #pagebreak()
% //
% // // == Parametric problem definition (design?)
% //
% // - find $k$
% //
% // - subject to
% //     - structural constraints
% //         - partial order on $k$ due to
% //             - fast/non fast reactions (TODO: as given by Reactome, but how?)
% //             #logic[
% //                 $
% //                     & forall r_f, r_s space.en ( r_f in R_"fast" and r_s in R_"slow" ) -> r_f > r_s
% //                 $
% //             ]
% //             - reaction modifiers (like above?)
% //     - for all dynamics of environment
% //         - avg concentration of species consistent to knowledge
% //
% //         $
% //             & exists t_0 space.en forall t space.en forall s \
% //             & quad (t > t_0 and s in S_"avg") -> s(t) in ["known range"]
% //         $


% \begin{definition}[Biological scenario]
%     A
% \end{definition} A scenario is defined by

   
% \begin{itemize}
%     \item a set of physical entities from which to start the expansions
%     \item a set ot pathways to which to limit (constraint?) reactions to
%     \item a max depth for recursion (/reachability) (TODO: a max depth in terms of nodes in the path, not in the number of reactions in the path, for that apoc is needed)
%
%     \item a set of physical entities to exclude from reachability
%
%     \item a partial order of the species 
%
% \end{itemize}
% (as per figure ... of UML etc...)


% \section{lll}

\newpage

\section{Constraint satisfiability problem}

% TODO: Version: 	4.3.6, Edition: 	enterprise

% TODO: biochemical network $\rightarrow$ biochemical network with dynamics
% $\rightarrow$ satisfiability problem $\rightarrow$ optimization problem

\begin{definition}[Biological model satisfiability problem]
    Given a biological model  $B = (G, \mathcal{K})$ let $\mathcal{S}$ be the
    set of concentrations of species of the species in the network,
    $\mathcal{S} = \{ s \; | \; s \in S \}$ and  $\mathcal{S}_\text{avg} =
    \{s_\text{avg} \; | \; s \in S\}$ the set of average concentrations of the
    species, $T \in \mathbb{R}$ the time horizon, the following constraints
    must hold:

    % \begin{itemize}
    \begin{equation}
         \forall s \in \mathcal{S} \quad 0 \leq s \leq 1
    \end{equation}
    \begin{equation}
        \begin{aligned}
            & \forall k_{r_1}, k_{r_2} \in \mathcal{K}, \; s \in S \\
            & \quad (s, r_1) \in E_\text{products} \land (s, r_2) \in E_\text{modifiers} \land r_1 \neq r_1 \Rightarrow k_{r_1} < k_{r_2}
        \end{aligned}
    \end{equation}
    \begin{equation}
        \forall s \in \mathcal{S}_\text{avg} \quad s(\phi \cdot T) - s(T) = 0
        % \Sum_{t \int [0, T]}
    \end{equation}

    % TODO: define what s at time "t" is
    % TODO: define domains somehow, and add an intermediary



    % \begin{itemize}
    %     \item $$
    %     % \item $\mathcal{S}_\text{avg} = \{s_\text{avg} \; | \; s \in S\}$
    %     % \item ll
    % \end{itemize}
\end{definition}

% #definition("constraint problem on a biological model")[
%     Given a model $G$ with target species $S_T$ and target pathways $P_T$ let
%     the following be a constraint problem
%
%
%     - $k : R -> RR^(|R|)$ // k is an assignement of the constants of the reaction, where |R| is the number of constants
%
%     *find* $k$
%     *subject to*
%     - partial order on $k$ from the structure of the graph
%     - partial order on the quantities
%     - constraint on enzymes such that
%         $
%             E + S <->^(k_1, k_(-1)) E S ->^(k_2) E + P
%         $
%     $
%         k_1, k_(-1) >> k_2
%     $
%
%     - for all dynamics of the environment
%         - average concentration of species consistent to knowledge
%         $
%             & exists t_0 space.en forall t space.en forall s \
%             & quad (t > t_0 and s in S_"avg") -> s(t) in ["known range"]
%         $
%
%     Environment: all possible cuts \
%     // DO NOT EXPAND OVER IT
%     we can have excluded species!
% ]
%
% \section{}

\section{Optimization problem}


\chapter{Blackbox optimization architecture}


% \begin{figure}
%     \noindent\makebox[\textwidth]{
%     \includegraphics[width=1.4\textwidth]{./images/orchestrator-worker.pdf}}
%     \caption{TEST} 
%     % \label{fig:reactome-uml}
% \end{figure}

% \begin{figure}
%     \noindent\makebox[\textwidth]{
%     \includegraphics[width=1.4\textwidth]{./images/orchestrator-worker.svg}}
%     \caption{TEST} 
%     % \label{fig:reactome-uml}
% \end{figure}


\section{Blackbox optimization}


Given a function $f : X \rightarrow Y$, which is expensive to compute, and an
optimization problem of the type $\argmin_{x \in X} f(x)$ is a blackbox optimization
problem if no information about the derivative of $f$ 

% When analyzing the scalability of a parallel algorithm on a HPC cluster, 
% an interesting problem is the one of trying to predict how would the 
% algorithm scale on a cluster with a higher degree of parallelism compared 
% to the one available for experiments.
% This document presents one possible way to make this kind of analysis 
% when the computation is asynchronous and the sequence of values in 
% the computation depends on the state of an orchestrator.
% OpenBox, a system design for generalized black-box optimization [1], will 
% be the main case study for this type of systems.

% TODO: optimization architecture

\section{OpenBox}

OpenBox is an efficient open-source system designed for solving gener­
alized black-box optimization (BBO) problems. It can be used either as 
a Standalone python package or Online BBO service \cite{li2021openbox}.

\section{Bayesian optimization}

\section{Orchestrator-worker infrastructure}

OpenBox is an efficient open-source system designed for solving generalized
black-box optimization (BBO) problems. It can be used either as a Standalone
python package or Online BBO service @open-box.

OpenBox has a great support for bayesian optimization, so that will be the main
subject of the analysis @open-box-automatic-algorithm-selection.

% Given a function $f: X -> Y$, which is expensive to compute, and an optimization
% problem of the type $"argmin"_(x in X) space f(x)$, the OpenBox service acts as
% an advisor which, when prompted, *suggests* the next point $x$ on which to
% compute the value $f(x)$. When a `worker` computes $f(x)$ it sends OpenBox an
% *observation*; each observation changes the state of OpenBox. When using the
% OpenBox as Service, the _worker needs to actively ask for the points_.

\section{Scalability analysis}

\chapter{Experiments}


\chapter{Appendix}

\appendix

\section{Neo4j Graph databse}

Neo4j is a native graph database, which means that it implements a true graph
model all the way down to the storage level. Instead of using a "graph
abstraction" on top of another technology, the data is stored in Neo4j in the
same way you may whiteboard your ideas.

https://neo4j.com/docs/getting-started/whats-neo4j/

% TODO: neo4j, how does it work, how does the apoc module work, 
% TODO: log some  possible scenario queries, do different max depths on it

\section{Reactome}

\begin{figure}
    \noindent\makebox[\textwidth]{
        \includegraphics[width=1.4\textwidth]{./images/reactome-uml.pdf}
    }
    % TODO built using the DatabaseObject reference
    \caption{UML model of a portion of the Reactome database} 
    \label{fig:reactome-uml}
\end{figure}

\begin{figure}
    \noindent\makebox[\textwidth]{
    \includegraphics[width=1.4\textwidth]{./images/biological-scenario-definition.pdf}}
    \label{fig:biological-scenario-definition-uml}
\end{figure}

\begin{figure}[H]
    \begin{center}
        \includegraphics[width=.7\linewidth]{./images/muscle-contraction-hierarchy.pdf}
    \end{center}
    \caption{Muscle contraction pathway hierarchy \cite{ReactomeStriatedMuscleContraction}}
    \label{fig:striated-muscle-contraction-hierarchy}
\end{figure}

\begin{figure}
    \begin{lstlisting}
    MATCH
    path =
        (reactionLikeEvent:ReactionLikeEvent)-[:catalystActivity]->
        (:CatalystActivity)-[:physicalEntity]->
        (physicalEntity:PhysicalEntity)
    WHERE
    EXISTS {
        MATCH (reactionLikeEvent)-[:input]->(physicalEntity)
    }
    RETURN COUNT(DISTINCT path)
    \end{lstlisting}
    \label{lst:modifiers-are-also-inputs}
\end{figure}

% \begin{center}
%     \includegraphics[width=\linewidth]{./images/biological-scenario-definition.pdf}
% \end{center}


% == CatalystActivity
%
% The role of #logic[PhysicalEntity] in #logic[_catalyst_activity_entity_] has
% multiplicity #logic[0..\*] because _"If a #logic[PhysicalEntity] can enable
% multiple molecular functions, a separate #logic[CatalystActivity] instance is
% created for each"_ #ref(<data-model-glossary>, supplement: [Page 5]).
%
% An additional constraint is required for active units, because _"If the
% #logic[PhysicalEntity] is a #logic[Complex] and a component of the complex
% mediates the molecular function, that component should be identified as the
% active unit of the #logic[CatalystActivity]."_ #ref(
%     <data-model-glossary>,
%     supplement: [Page 5],
% )
%
% // TODO: does it expand to multiple level complexes?
%
% #constraint(
%     highlight(`C.CatalystActivity.active_unit_is_component_of_complex`),
%     ```
%     forall catalyst_activity, complex, complex_component
%         (
%             CatalystActivity(catalyst_activity) and
%             Complex(complex) and
%             PhysicalEntity(complex_component) and
%             catalyst_activity_entity(catalyst_activity, complex) and
%             catalyst_activity_active_unit(
%                 catalyst_activity,
%                 complex_component
%             )
%         ) ->
%             complex_has_component_entity(complex, complex_component)
%     ```,
% )
%
% == Compartment
%
% // #TODO[move this information to the #logic[_compartment_entity_] association, or
% //     to #logic[PreferredCompartmentForSimulation]]
%
% The #logic[Compartment] class has some quirks. In Reactome, the
% #logic[Compartment]'s role in the #logic[_compartment_entity_] association has
% multiplicity #logic[0..\*]. The problem is that the SBML model requires
% #logic[1..1] multiplicity for this association to be simulated.
%
% In Reactome there are currently (TODO: version??) 19 physical entities which
% don't have a compartment (see queries/helper.cypher), so this can be easily
% solved by just adding a *default compartment* to the SBML model to which these
% entities map to.
%
% On the other hand there are 14046 entities which have multiple compartments
% (TODO: how many compartments has each exactly?), so the easiest choice right now
% is to just pick any of them. For this reason the
%
%
% // TODO: actual model requires 1..1 compartments for PhysicalEntity
% //
% // - if no compartment is present just use a default one
% // - if multiple compartments are present use whichever you want (for now)
%
% // == DatabaseObjectWithStableId
% //
% // #constraint(
% //     [C.DatabaseObjectWithStableId.either_database_id_or_id_is_defined],
% //     ```
% //     forall object
% //         DatabaseObjectWithStableId(object) ->
% //             exists id
% //                 database_id(object, id) or id(object, id)
% //     ```,
% // )
%
% == Pathway
%
% The instances of #logic[Pathway] are organized hierarchically, i.e. all the
% signaling pathways are collected under the Signal Transduction top level
% #logic[Pathway] (#logic[StableIdVersion] R-HSA-162582.13). This allows to easily
% extract a subset of reactions by specifying the _target pathways_ in a model and
% taking into consideration only the reactions which are included, both directly
% or indirectly, in that pathway (see the #logic[included_reactions()] operation).
%
% There are about 34 top level pathways.
%
% #operation(
%     `included_reactions`,
%     type: `ReactionLikeEvent [0..*]`,
%     // prec: ```
%     // ```,
%     postconditions: ```
%     result =
%         { reaction |
%             ReactionLikeEvent(reaction) and
%             pathway_has_event(this, reaction) }
%         $union$
%         { reaction | exists pathway
%             Pathway(pathway) and
%             pathway_has_event(this, pathway) and
%             #included_reactions(pathway, reaction) }
%     ```,
% )


% TODO: what is the performance of the query (depths: 2, 4, 8, 16, restart each time)
% TODO: how big does it need to be








\section{SBML simulation with Roadrunner}



\section{HPC Cluster emulation with Docker Compose}


% = OpenBox on the Slurm Workload Manager
%
% The HPC cluster at the Computer Science Department has some restrictions in
% place, as it's used by many different teams / students and no single user can
% request the indefinite usage of the whole cluster for a single job (jobs have a
% time limit of 6, 24 or 72 hours based on permissions and resoruces required).
%
% // TODO: somehow store the OpenBox model of a job
% // TODO: test both "students" and "multicore" partition to see which is more convenient
% The goals of this section are to
% - be able to run an OpenBox throught *multiple sessions*
% - run *multiple smaller jobs* to increase *fairness* among users, instead of
%     running a single big job for the whole simulation
% - provide a simple framework that can be used *locally to simulate* executions
%     on the cluster
%
% == Analysis
%
% In order to use OpenBox on the cluster in different sessions, it's a good idea
% to store the results of the simulations in a database (i.e. PostgreSQL) to
% retrieve the data of different session for an overall analysis.
%
% // #page(
% //     flipped: true,
% //     width: auto,
% //     height: auto,
% //     align(center + horizon, image("docs-biological-scenario-definition.svg")),
% // )
% #box(inset: (y: 5pt), align(center, image("./docs-openbox-jobs.svg")))
%
% === Data types specification
%
% #logic[
%     `SlurmJobId = Integer >= 1` \
%     `String1 = String` matching regex ```js /^\S$|^\S.*\S$/```
% ]
%
% #pagebreak()
%
% === Classes specification
%
% ==== Job
%
% #constraint(
%     highlight(`C.Job.all_parameters_are_instantiated`),
%     ```
%     forall job, blackbox, parameter
%       (
%         Job(job) and
%         Blackbox(blackbox) and
%         Parameter(parameter) and
%         blackbox_job(blackbox, job) and
%         blackbox_parameter(blackbox, parameter)
%       ) ->
%         job_parameter(job, parameter)
%     ```,
% )
%
% #constraint(
%     highlight(`C.Job.continuity_1`),
%     ```
%     forall job, submit_time, start_time
%       (
%         Job(job) and
%         #submit_time(job, submit_time) and
%         #start_time(job, start_time)
%       ) ->
%         submit_time <= start_time
%     ```,
% )
%
% #constraint(
%     highlight(`C.Job.continuity_2`),
%     ```
%     forall job, start_time, end_time
%       (
%         Job(job) and
%         #start_time(job, start_time) and
%         #end_time(job, end_time)
%       ) ->
%         start_time <= end_time
%     ```,
% )
%
% #pagebreak()
%
% == Implementation
%
% Diagram restructuration for PostgreSQL. The SQL code is available in the
% `migration.sql` file.
%
% #box(inset: (y: 5pt), align(
%     center,
%     image("./docs-openbox-jobs-restructuration.svg"),
% ))
%
% === Data types definitions
%
% // TODO: I could just use a serial id, but it is nice to have the job_id that can be generated only by slurm, so one can't add a "fake job"
% ```sql
% CREATE DOMAIN String1 AS varchar CHECK(value ~ '^\S$|^\S.*\S$');
% CREATE DOMAIN SlurmJobId AS integer CHECK(value >= 1);
% ```
%
% === Additional constraints
%
% ==== Job
%
% #constraint(
%     highlight(`C.Job.end_implies_job_was_scheduled`),
%     ```
%     forall job, end_time
%       (Job(job) and #end_time(job, end_time)) ->
%         exists start_time #start_time(job, start_time)
%     ```,
% )
%
% A result is present if and only if the job ended
%
% #constraint(
%     highlight(`C.Job.result_only_on_end_time`),
%     ```
%     forall job, job_result
%       (Job(job) and #result(job, job_result)) ->
%         exists end_time #end_time(job, end_time)
%     ```,
% )
%
% #constraint(
%     highlight(`C.Job.end_time_only_on_result`),
%     ```
%     forall job, end_time
%       (Job(job) and #end_time(job, end_time)) ->
%         exists job_result #result(job, job_result)
%     ```,
% )
%
% #page(bibliography("bibliography.bib"))









\newpage

% % 268 words circa for introduction
%
% % TODO: check paper on boolean networks for stationary states
% Steady states and attractors
%
% One consequence of evolution is the similarity of bio­
% logical organisms of different species.
%
% allows for the use of model organisms and for the critical
% transfer of insights gained from one cell type to other cell
% types. Applications include, for example, prediction of
% protein function from similarity, prediction of network
% properties from optimality principles, reconstruction of
% phylogenetic trees, or the identification of regulatory
% DNA
%
% % TODO: we add 
%
% If we observe biological phenomena, we are confronted
% with various complex processes that often cannot be
% explained from first principles and the outcome of which
% cannot reliably be foreseen from intuition. Even if general
% biochemical principles are well established (e.g., the
% central dogma of transcription and translation or the bio­
% chemistry of enzyme-catalyzed reactions), the bio­ chemistry of individual molecules and systems is often
% unknown and can vary considerably between species.
% Experiments lead to biological hypotheses about individ­
% ual processes, but it often remains unclear whether these
% hypotheses can be combined into a larger coherent pic­
% ture because it is often difficult to foresee the global
% behavior of a complex system from knowledge of its
% parts. Mathematical modeling and computer simulations
% can help us to understand the internal nature and
% dynamics of these processes and to arrive at predictions
% about their future development and the effect of interac­
% tions with the environment.
%
% In a broad sense, a model is an
% abstract representation of objects or processes that
% explains features of these objects or processes
%
% On the structure of biological systems
%
% % TODO: Having a model which is stable can help in studing situations in diseases etc...
% % TODO: generally we want to find steady states, cite boolean networks paper
% % TODO: this search explodes, i.e. steady states in boolean networks
% % TODO: find a paper that says it's very hard to find assignemnts for variables etc.. (cite the same ones in the SMC paper, or something) 
%
% Therefore, stochastic algorithms must be used instead. Statistical
% Model Checking (SMC) (Larsen and Legay, 2016) offers a wide range
% of frameworks and probabilistic methods to analyze and verify large
% models of complex systems. While these tools generally cannot provide
% definite answers to verification problems, they can actually be applied
% to real-world use-cases and generate approximations of Key Perfor-
% mance Indicators (KPIs)
%
%
% % The growing need for tools designed for the formal verification of
% % complex systems faces a fundamental problem of scalability
%
% \begin{itemize}
%     \item Solution algorithms and computer programs can be used independently of the concrete system. Modeling is cheap compared with experiments.
%     \item Modeling can assist experimentation. With an adequate model, one may test different scenarios that are not accessible by experiment.
%     \item One may follow time courses of compounds that cannot be measured in an experi ment. 
%     \item One may impose perturbations that are not feasi­ ble in the real system.
%     \item Model simulations can be repeated often and for many different conditions.
%     \item Model results can often be presented in precise mathe matical terms that allow for generalization. 
%     \item Finally, modeling allows for making well-founded and testable predictions.
% \end{itemize}
%
% % \begin{itemize}
% %     % \item Modeling drives conceptual clarification. It requires verbal hypotheses to be made specific and conceptually rigorous.
% %     % \item Modeling highlights gaps in knowledge or understanding. During the process of model formulation, unspecified components or interactions have to be determined.
% %     \item Modeling provides independence of the modeled object. 
% %     \item Time and space may be stretched or compressed ad libitum.
% %     \item Models exert by themselves no harm on animals or plants and help to reduce ethical problems in experi­ ments. They do not pollute the environment.
% %     \item One may cause precise perturbations without directly changing other system components, which is usually impossible in real sys­ tems.
% %     \item Model simulations can be repeated often and for many different conditions.
% %     \item Model results can often be presented in precise mathe matical terms that allow for generalization. 
% %     \item Graphical representation and visualization make it easier to understand the system.
% %     \item Finally, modeling allows for making well-founded and testable predictions.
% % \end{itemize}
%
% % \section{Computational modeling of biological systems} % TODO: Motivation
% %
% % % augmenting / quantifying
% % \section{Qualitative network models augmentation} % TODO: Contribution
% %
% % \subsection{Qualitative network models}
% % TODO: here cite reactome, put in "notes" the UML used later for the query etc...
% %
% % \section{Outline} % TODO: Outline
% %
% %
% % % This article is organized as follows. Section 2 presents preliminaries
% % % and discusses state-of-the-art SAs, with a special emphasis on EBStop
% % % and AA. Section 3 defines the formal framework used throughout the
% % % article. Section 4 and Section 5 describe EAA and our massively parallel
% % % implementation, respectively. Section 6 is devoted to the experimental
% % % analysis of our algorithm and parallel tool. Finally, Section 7 draws
% % % conclusions and perspectives
% %
% % \begin{algorithm}
% % \caption{Model generation}
% % \begin{algorithmic}
% % \Require $n \geq 0$
% % \Ensure $y = x^n$
% % \State $i \gets 10$
% % \If{$i\geq 5$} 
% %     \State $i \gets i-1$
% % \Else
% %     \If{$i\leq 3$}
% %         \State $i \gets i+2$
% %     \EndIf
% % \EndIf 
% % \end{algorithmic}
% % \end{algorithm}
% %
% %
% % \chapter{Preliminaries}
% %
% % We denote by $\mathbb{R}, \mathbb{R}_{0_{+}}, \mathbb{Z}, \mathbb{N}$ the sets of, respectively, real, non-
% % negative real, integer, and non-negative integer numbers.
% %
% % $\theta \in [-20, 20]^{k}$
% %
% %
% % % \section{On the usefulness of }
% %
% % \section{Computational modeling of biological systems}
% %
% % TODO: what is the general problem I want to solve?
% %
% % % \section{Qualitative biological networks}
% %
% % Given a set of target species, a set of constraints on the target species 
% % (constraints which model a scenario that could present, for example, in a 
% % disease) and by taking into account all the reactions within a set target 
% % pathways that lead to the production, both directly and indirectly, of the 
% % target species, the goal is to find a subset of virtual patients for the described 
% % scenario.
% %
% %
% % \begin{definition}[Biological network] A biological network $G$ is a tuple $(S, R, E, \sigma)$ where
% % \begin{itemize}
% %     \item $S = U \sqcup X \sqcup Y$ is the set of species of the biological network, where
% %         \begin{itemize}
% %             \item $U$ is the set of input species
% %             \item $X$ is the set of other species in the network
% %             \item $Y$ is the set of output species
% %         \end{itemize}
% %     \item $R$ is the set of reactions in the biological network
% %     \item $E \subseteq S \times R$
% % \end{itemize}
% % \end{definition}
% %
% % \begin{definition}[Optimization problem]
% % \end{definition}
% %
% % % \section{Qualitative}
% %
% % \chapter{Quantitative model generation}
% %
% % \chapter{Constraint satisfiability problem}
% %
% % \chapter{Parallelization}
% %
% % \chapter{Scalability analysis}
% %
% % \chapter{Experiments}
% %
%
% % \section{}
%
% % \medskip
%
% jkjk
% kjjlkjlk
% kjlkjlkj
%
\printbibliography
%
% "Don’t be overwise; fling yourself straight into life, without deliberation;
% don’t be afraid - the flood will bear you to the bank and set you safe on
% your feet again." Fyodor Dostoevsky

% \lettrine[lines=2, findent=3pt, nindent=0pt]{I}{}n the field of probabilistic  password guessing models...
%
% \bigskip
% In \hyperref[chap:1]{Chapter~\ref*{chap:1}} we  briefly present...
%
% \bigskip
% In \hyperref[chap:2]{Chapter~\ref*{chap:2}} we summarize...
% \bigskip
% In \hyperref[chap:3]{Chapter~\ref*{chap:3}} we  briefly present...
%
% \bigskip
% In \hyperref[chap:4]{Chapter~\ref*{chap:4}} we summarize...
% \bigskip
% In \hyperref[chap:5]{Chapter~\ref*{chap:5}} we  briefly present...
%
% \chapter{Background and Related Work}

\end{document}

% Boolean networks is a well-established formalism for modelling biological systems. A vital challenge for analyzing a Boolean network is to identify all the attractors.
% This becomes more challenging for large asynchronous Boolean networks, due to the asynchronous scheme.
% Existing methods are prohibited due to the well-known state-space explosion problem in large Boolean networks. 
% In this paper, we tackle this challenge by proposing a SCC-based decomposition method. We prove the correctness of our proposed method and demonstrate its efficiency with two real-life biological networks.
%
%
% - ME VERSION: We want to find the attractors, in large asynchronous networks is hard (State Space explosion, nice), so we decompose the network in SCC (Strongly Connected Components) 


% TODO: On ATTRACTORS

% A *GRN* _(Genere Regulatory Network)_ is a collection of molecular regulators that interact with each other and with other substances in the cell to govern the gene expression levels of mRNA and proteins which, in turn, determine the function of the cell.
%
% A *molecular regulator*  is a molecule that controls the activity, expression, or function of other molecules or processes within a cell or organism.
%
% An attractor of a biological system is a set of the system’s states satisfying that any two states in this set can be reached from each other and the system remains in this set until some external stimulus pushes the system out of it.
%
% Attractors are hypothesised to characterise cellular phenotypes or to correspond to functional cellular states such as 
% - proliferation (the process of cell growth and division, resulting in an increase in the number of cells)
% - apoptosis (the process of programmed cell death, in which a cell initiates self-destructions; necrosis is a 'messy' uncontrolled cell death)
% - differentiation (the process where a less specialized cell develops into a more specialized cell type, acquiring distinct characteristics and functions)
% - homeostasis (a finely tuned balancing act involving the interplay of various cellular processes such as oxidative phosphorylation, autophagy protein ubiquitination, and sumoylation. These pathways work in harmony to ensure cells function optimally, adapting to changes and stressors)


% TODO: on attractor detection

% 1. Enumeration
%     - In the early 2000s, an enumeration and simulation method has been proposed. 
%     - The attractors are detected by running simulation from each of the possible states. 
%     - This method is largely restricted by the network size since the required computational time grows exponentially with the number of nodes.
% 2. Irons 
%     - 2006 method to handle BNs with maximum 50 nodes
% 3. BDDs (Binary Decision Diagrams)
%     - Basically decision trees
%     - It has efficient operations to compute forward and backward reachable states
%     - To identify attractos it's enough to identify the fix point set in the corrisponding transition system
% 4. SAT solvers
%     - The transition relation of the BN is unfolded for a bounded number of steps, and it's presented as a propositional formula
%     - The process is repeated for bigger bounds until no valid path can befound
%     - It depdends on the number of nodes in the BN and the number of unfolding steps required
% 5. BNs decomposition
%     - The main idea is to decmpose a large BN into small components based on its structure, detect attractors in the small compoments, and then recover the attractors of the original BN 
% 6. other methods for large networks (cannot guarantee to find all attractors)
%     - approximation methods
%     - reduction methods
%     - methods designed for specific types of attractors



% An attractor of a BN is a set of states satisfying that any state in the set can be reached from any other state in this set and no state in this set can reach any other state that is not in this set.
%
% Attractors are hypothesised to characterise cellular phenotypes.
%
% We refer to an attractor togeher with its state transition relation as an attractor system (like an STS).
% - its state are called attractor states
% - it is not guaranteed that an asynchronous attractor is a loop
% - in asynchronous systems the attractor might inculde serveral loops
%
% Types of attractors:
% - singleton attractor (a selfloop)
% - simple loop
% - complex loop

% TODO:
% # What question is the model supposed to answer?
% # Is it built to explain a surprising observation?
% # Is it built to relate separate observations with each other and with previous knowledge?
% # Is it built to make predictions, for example, about the effect of specific perturbations?



% - proteomics: studio di tutte le proteine prodotte o modificate da un organismo, sistema biologico o cellula in un determinato momento
% - REACTOME: Neo4j, graph database, Cypher
% - REACTOME: [MySQL dump](https://reactome.org/download-data)
% - [REACTOM BNs](https://reactome.org/docs/training/ReactomeFIVizapp.pdf#:~:text=The%20main%20features%20of%20the%20plug%2Din%20are,edge%20in%20the%20the%20network%20view%20panel.&text=The%20main%20features%20of%20the%20plug%2Din%20are,edge%20in%20the%20the%20network%20view%20panel.)
% - BIOMARCATORE ONCOLOGICO: un qualche tipo di sostanza che può essere rilevata ed è associata alla presenza / crescita di un tumore 
% - MENSTRUAL BLEEDING: excessively havey flow, abnormal bleeding
% - PATHWAY ENDOCRINOLOGICO: si riferisce al sistema di comunicazione e regolazione all'interno del corpo che coinvolge gli ormoni, messaggeri chimici prodotti dalle ghiandole endocrine. Questi ormoni viaggiano nel sangue verso le cellule bersaglio, dove interagiscono con recettori specifici per influenzare le funzioni cellulari
% - NEFROLOGICO: è la branca della medicina che si occupa dello studio, della diagnosi e del trattamento delle malattie dei reni e delle vie urinarie
% - ONCOLOGICO: si riferisce a tutto ciò che è relativo ai tumori, sia benigni che maligni, e alla branca della medicina che li studia, l'oncologia

% TODO: NOTE2

% vincoli
%     - concentrazione media specie più o meno quella nota
%     - vincoli di ordinamento per cui una reazione è più veloce di un'altra è soddisfatta
%     - per tutti gli scenari di input ai morsetti
%     - system model checking
%
%
% - DONE: articolo su statistical model checking
% - TODO: fast Reaction? How to check if reaciton is fast?
% - TODO: modificatori, catalisi? Come capire quali sono?
% - TODO: isProduct invece di output? (indice opposto)
% - TODO: ratio degli scenari in cui finisce
%
% - NONE! TODO: apricopt add nevergrad + OpenAI
% - TODO: casi particolari tPA (PLAT)
% - riccardo curcio OpenAI tool
% - NOMAD: mesh semplificata del problem
%
% - NONE! TODO: estendiamo apricopt (ottimizzatore black box OpenAI)
% - TODO:
%     - ottimizzatore BBO fa search K
%         - vincoli strutturali (di dominio)
%         - la probabilità K fa fallire X < epsilon
%     - Env + modello SBML simulatore
%     - la BBO del simulatore è il tuo EA
%
% - NOMAD ad ogni chiamata del BBO vuole sia i vincoli strutturali sia simulabili
%
% - NONE! TODO: aggiungere tool BBO a docker-ai-tools
% https://en.wikipedia.org/wiki/Law_of_mass_action
%
% # note
%
% - Estendere EAA con nuovo algoritmo sistem model checking (estendi)
% - Nuovo ottimizzatore OpenAI su apricopt
%
%
%
% # note 2
%
% - Generare modello SBML da Reactome
% - simulatore AMICI (+ apricopt se serve)
%
% VEDERE SE vincolo 2 rf < rs sia così
%
% - tenere modello in RAM e simulare tante volte cambiando parametri
% - biouml
% - vettore parametri
% - vettore traiettorie
% - multiprocess Pool Python
%
% TODO:
% - isola reactome (sceglie isola + stacca frontiera)
% - generare modello SBML quantitativo eq. diff. parametrico nelle costanti
%
% - template che genera l'equazione sia modificabile (law of mass action)
% - utente, quando vuole generare il modello, può decidere quale template usare
% - idealmente alcune reazioni modellate con una certa legge, e altre reazioni con altre (ma sembra inverosimile che per ogni reazione si conosce il template più opportuno)
% - template oggetti di una classe, data una reazione qualitativa, genera un'equazione differenziale parametrica nelle costanti cinetiche (velocità)

% TODO: this is a macroscopical + phenomenological approach (compared to microscopical simulation of single components)


% TODO: https://sbgn.github.io/ for images
% TODO: MIRIAM could be useful, it defines standards for different types of systems biology models
% TODO: the hierarhical structure of reactome, use here, describe in appendix
